src/Microsoft.DotNet.Configurer/NuGetCacheSentinel.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Extensions.EnvironmentAbstractions;
+using Microsoft.DotNet.ProjectModel.Resolution;
+
+namespace Microsoft.DotNet.Configurer
+{
+    public class NuGetCacheSentinel : INuGetCacheSentinel
+    {
+        public static readonly string SENTINEL = $"{Product.Version}.dotnetSentinel";
+
+        private readonly IFile _file;
+
+        private string _nugetCachePath;
+
+        private string NuGetCachePath
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(_nugetCachePath))
+                {
+                    _nugetCachePath = PackageDependencyProvider.ResolvePackagesPath(null, null);
+                }
+
+                return _nugetCachePath;
+            }
+        }
+
+        private string Sentinel => Path.Combine(NuGetCachePath, SENTINEL);
+
+        public NuGetCacheSentinel() : this(string.Empty, FileSystemWrapper.Default.File)
+        {
+        }
+
+        internal NuGetCacheSentinel(string nugetCachePath, IFile file)
+        {
+            _file = file;
+            _nugetCachePath = nugetCachePath;
+        }
+
+        public bool Exists()
+        {
+            return _file.Exists(Sentinel);
+        }
+
+        public void CreateIfNotExists()
+        {
+            if (!Exists())
+            {
+                _file.CreateEmptyFile(Sentinel);
```

Comment:
> NUGET_PACKAGES env variable to some non-writable location.
We should indeed make sure we don't do terrible things if we can't make this process work.  Ideally we fail gracefully early before doing anything expensive.

Here's a suggestion for the sentinel. Start by checking for the sentinel: S.  If it exists skip.  If not, check for S.inprogress and skip|return|wait.  Create a new file with name S.inprogress with delete on close and NO sharing and keep it open.  If you can't create it bail without doing anything.  Once your done write S and close S.inprogress (deleting it).  Only problem is somehow someone created S.inprogress and orphaned it, you can detect that by opening it.  If you can open it it means that no other process was running (since we had the handle with no sharing) and you can treat it as your turn to do the first run.